client/tinychain/ml/linalg.py

```diff
@@ -274,3 +277,81 @@ def svd(matrix: Tensor) -> Tuple:
     """Return the singular value decomposition of the given `matrix`."""
 
     raise NotImplementedError("singular value decomposition")
+
+
+def householder_bidiag(U: Tensor, W: Tensor, e: Tensor):
+    """Householder's reduction to bidiagonal form"""
+    @post_op
+    def check_rows_num(cxt, U: Tensor, scale: F32, i: UInt, m: UInt, n: UInt, l: UInt):
+        scale = F32(tc.linalg.norm(U[i:, i]))**2
+
+        # @post_op
+        def change_rows(cxt, U: Tensor, scale: F32, i: UInt, l: UInt, n: UInt):
+            s = F32(After(U[i:, i].write(U[i:, i] / scale),
+                    F32(tc.linalg.norm(U[i:, i]))))
+            f = F32(U[i, i])
+            g = F32(If(f >= F32(0), -s, s))*1
+            h = F32(f * g - s)
+            U[i, i].write(f - g)
+            cxt.h = h
+
+            @closure
+            @post_op
+            def step(j: UInt, U: Tensor) -> Map:
+                return Map(After(
+                    U[i:, j].write(U[i:, j] + U[i:, i] *
+                                   F32(F32(U[i:, i].mul(U[i:, j]).sum()) / cxt.h)),
+                    Map(U=U, j=j + 1)))
+
+            @closure
+            @post_op
+            def cond(j: UInt, U: Tensor):
+                return j < n
+
+            result = Map(While(cond, step, Map(
+                U=U.copy(),
+                j=l
+            )))
+
+            U = Tensor(result['U'])
+            U[i:, i].write(U[i:, i] * scale)
+
+            return Tuple([U, scale, g])
+
+        return If((i >= m).logical_or(scale <= EPS), Tuple([U, scale, 0.0]), change_rows(cxt, U=U, scale=scale, i=i, l=l, n=n))
+
+    @post_op
+    def check_cols_num(cxt, U:Tensor, scale: F32, i: UInt, m: UInt, n:UInt, l: UInt, e: Tensor):
+        scale = F32(tc.linalg.norm(U[i, l:]))**2
+    
+        # @post_op
+        def change_cols(cxt, U: Tensor, scale: F32, i: UInt, l: UInt, n: UInt, e: Tensor):
+            s = F32(After(U[i, l:].write(U[i, l:] / scale), F32(tc.linalg.norm(U[i, l:]))))
+            f = F32(U[i, l])
+            g = F32(If(f >= F32(0), -s, s))*1
+            h = F32(f * g - s)
+            U[i,l].write(f - g)
+            e[l:].write(U[i,l:] / h)
+
+            @closure
+            @post_op
+            def step(j: UInt, U: Tensor) -> Map:
+                return Map(After(
```

Comment:
nit: the return annotation will make sure that the calling code gets a `Map` back, you don't need to call `Map` here